enhanced-stack-trace-queries.md

```diff
@@ -0,0 +1,111 @@
+# Enhanced stack trace queries
+
+## Backround
+
+The SDKs have a built-in `__stack_trace` query that can be used display where the workflow is blocked at the time of the
+query (the last processed workflow task).
+
+The response to a `__stack_trace` query is a single string representing a list of stack individual stack traces
+separated by double newlines.
+
+Example from the TS SDK:
+
+```
+    at someInternalFunction (/path/to/workflows/my-workflow.ts:18:10)
+    at someWorkflowFunction (/path/to/workflows/my-workflow.ts:60:20)
+
+    at someInternalFunction2 (/path/to/workflows/my-workflow.ts:100:30)
+    at someWorkflowFunction2 (/path/to/workflows/my-workflow.ts:123:5)
+```
+
+While this is valuable for developers to be able to understand how their code correlates to event histories, it is quite
+limited since it requires looking up the actual code to get the full picture.
+
+## Proposal
+
+SDKs can optionally implement an `__enhanced_stack_trace` query (name TBD) where instead of returning the stack traces
+as a single string, they return a pre-defined structure, that structure will also include file sources.
+
+The web UI can use this response to show the workflow code and highlight locations where it is blocked.
+
+### Proposed structure
+
+```ts
+export interface SDKInfo {
+  name: string;
+  version: string;
+}
+
+/**
+ * Represents a slice of a file starting at lineOffset
+ */
+export interface FileSlice {
+  content: string;
+  lineOffset: number;
+}
+
+/**
+ * A pointer to a location in a file
+ */
+export interface FileLocation {
+  /**
+   * Path to source file (absolute or relative).
+   * When using a relative path, make sure all paths are relative to the same root.
+   */
+  filepath: string;
+  line: number;
+  column: number;
```

Comment:
In that case the UI would need to degrade to something like the current stack trace view. Maybe the SDK should detect this and return a NotImplemented error back to the UI which would fallback to string based stack trace.

Reply:
I think just making line/column optional is fine. In Java and others when compiling without debug options, you'll get method and what not with no line numbers (I forget whether file is absent).

Reply:
I'm adding:
If worker cannot get the required information for enhancing the stack trace, it will still respond with a structured stack trace but sources and some location fields might be missing.

I'll also make it the worker's responsibility to specify which fields are going to be included in the response.
Something like:
```ts
interface Capabilities {
  lineNumbers?: boolean;
  sources?: boolean;
  ...
}
```

It's more info but makes it easier for the UI to pick the right visualization.

Reply:
Why do that instead of simply using the presence or absence of those things to determine whether or not to render them? All the information is there already, this is just an opportunity to accidentally incorrectly construct the types.

Reply:
> Why do that instead of simply using the presence or absence of those things to determine whether or not to render them? All the information is there already, this is just an opportunity to accidentally incorrectly construct the types.

I’d also add that some information might be available on some files/frame and not on other. Flags at top level might not help much in such cases…

Reply:
SGTM, you both make a good point